File: ds/Queues/queues-class.cpp
#include<iostream>

class Node {
  public:
    int data;
    Node *next = nullptr;

    Node (int val) {
      data = val;
    }
};

class Queue {
  public:
    Node *first = nullptr, *last = nullptr;

    void enqueue (Node *node) {
      if (first == nullptr) {
        first = node;
        last = node;
      }
      else {
        last->next = node;
        last = last->next;
      }
    }

    Node *dequeue () {
      if (first == nullptr) {
        return nullptr;
      }
      else {
        Node *removeElem = first;
        first = first->next;
        return removeElem;
      }
    }

    void printQueue () {
      Node *start = first;
      while (start) {
        std::cout<<start->data<<std::endl;
        start = start->next;
      }
    }
};

int main () {
  Node root(1);
  Node two(2);
  Node three(3);

  Queue s;
  s.enqueue(&root);
  s.enqueue(&two);
  s.enqueue(&three);

  s.printQueue();

  s.dequeue();
  s.dequeue();
  s.dequeue();
  s.dequeue();

  s.printQueue();
}

File: ds/Queues/queues-class.js
class QueueNode {
  constructor (val) {
    this.value = val;
    this.next = null;
    this.prev = null;
  }
}

class Queue {
  constructor () {
    this.head = null;
    this.tail = null;
  }

  enqueue (val) {
    if (this.head == null) {
      this.head = new QueueNode(val);
      this.tail = this.head;
    }
    else {
      let newNode = new QueueNode(val);
      this.head.prev = newNode;
      newNode.next = this.head;
      this.head = newNode;
    }
  }

  dequeue () {
    if (this.head = null) {
      console.log('The queue is empty');
      return false;
    }
    else {
      let returnVal = this.tail.value;
      this.tail = this.tail.prev;
      return returnVal;
    }
  }

  print () {
    let temp = this.tail;

    while (temp) {
      console.log(temp.value);
      temp = temp.prev;
    }
  }
}

let q = new Queue();

q.enqueue(1);
q.enqueue(2);
q.print();
q.dequeue();
q.print();